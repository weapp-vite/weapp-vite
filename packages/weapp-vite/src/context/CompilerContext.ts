import type { AutoImportService } from '../runtime/autoImportPlugin'
import type { BuildService } from '../runtime/buildPlugin'
import type { ConfigService } from '../runtime/config/types'
import type { JsonService } from '../runtime/jsonPlugin'
import type { NpmService } from '../runtime/npmPlugin'
import type { RuntimeState } from '../runtime/runtimeState'
import type { ScanService } from '../runtime/scanPlugin'
import type { WatcherService } from '../runtime/watcherPlugin'
import type { WxmlService } from '../runtime/wxmlPlugin'
import '../config'

export interface CompilerContext {
  runtimeState: RuntimeState
  configService: ConfigService
  npmService: NpmService
  wxmlService: WxmlService
  jsonService: JsonService
  watcherService: WatcherService
  autoImportService: AutoImportService
  buildService: BuildService
  scanService: ScanService
}

export type MutableCompilerContext = Partial<Omit<CompilerContext, 'runtimeState'>> & {
  runtimeState: RuntimeState
  configService?: ConfigService
  npmService?: NpmService
  wxmlService?: WxmlService
  jsonService?: JsonService
  watcherService?: WatcherService
  autoImportService?: AutoImportService
  buildService?: BuildService
  scanService?: ScanService
}
