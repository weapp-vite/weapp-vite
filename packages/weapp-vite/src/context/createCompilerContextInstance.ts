import type { CompilerContext, MutableCompilerContext } from './CompilerContext'
import { createAutoImportServicePlugin } from '../runtime/autoImportPlugin'
import { createBuildServicePlugin } from '../runtime/buildPlugin'
import { createConfigServicePlugin } from '../runtime/configPlugin'
import { createJsonServicePlugin } from '../runtime/jsonPlugin'
import { createNpmServicePlugin } from '../runtime/npmPlugin'
import { createRuntimeState } from '../runtime/runtimeState'
import { createScanServicePlugin } from '../runtime/scanPlugin'
import { createWatcherServicePlugin } from '../runtime/watcherPlugin'
import { createWxmlServicePlugin } from '../runtime/wxmlPlugin'

export function createCompilerContextInstance(): CompilerContext {
  const context = {
    runtimeState: createRuntimeState(),
  } as MutableCompilerContext

  const runtimePlugins = [
    createConfigServicePlugin(context),
    createWatcherServicePlugin(context),
    createWxmlServicePlugin(context),
    createJsonServicePlugin(context),
    createScanServicePlugin(context),
    createAutoImportServicePlugin(context),
    createNpmServicePlugin(context),
    createBuildServicePlugin(context),
  ]

  Object.defineProperty(context, Symbol.for('weapp-runtime:plugins'), {
    configurable: false,
    enumerable: false,
    writable: false,
    value: runtimePlugins,
  })

  return context as CompilerContext
}
