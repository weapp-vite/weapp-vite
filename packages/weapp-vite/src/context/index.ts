export type { AutoImportService } from '../runtime/autoImportPlugin'
export type { BuildOptions, BuildService } from '../runtime/buildPlugin'
export type { ConfigService, LoadConfigOptions, LoadConfigResult } from '../runtime/config/types'
export type { JsonService } from '../runtime/jsonPlugin'
export type { NpmService } from '../runtime/npmPlugin'
export type { RuntimeState } from '../runtime/runtimeState'
export type { ScanService } from '../runtime/scanPlugin'
export type { WatcherService } from '../runtime/watcherPlugin'
export type { WxmlService } from '../runtime/wxmlPlugin'
export * from './CompilerContext'
export * from './getInstance'
