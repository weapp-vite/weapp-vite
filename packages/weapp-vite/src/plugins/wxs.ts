import type { EmittedFile, PluginContext } from 'rolldown'
import type { Plugin } from 'vite'
import type { CompilerContext } from '../context'
import type { WxmlDep } from '../types'
import { removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import { LRUCache } from 'lru-cache'
import path from 'pathe'
import { jsExtensions } from '../constants'
import { transformWxsCode } from '../wxs'

export const wxsCodeCache = new LRUCache<string, string>(
  {
    max: 512,
  },
)

export function wxs({ configService, wxmlService }: CompilerContext): Plugin[] {
  const wxsMap = new Map<string, {
    emittedFile: EmittedFile
  }>()

  async function transformWxs(this: PluginContext, {
    wxsPath,
  }: {
    wxsPath: string
  }) {
    this.addWatchFile(wxsPath)
    if (await fs.exists(wxsPath)) {
      const arr = wxsPath.match(/\.wxs(\.[jt]s)?$/)
      let isRaw = true
      if (arr) {
        isRaw = !arr[1]
      }
      const rawCode = await fs.readFile(wxsPath, 'utf8')
      let code = wxsCodeCache.get(rawCode)

      const dirname = path.dirname(wxsPath)
      if (code === undefined) {
        const { result, importees } = transformWxsCode(rawCode, {
          filename: wxsPath,
        })
        if (typeof result?.code === 'string') {
          code = result.code
        }
        await Promise.all(
          importees.map(({ source }) => {
            return transformWxs.call(this, {
              wxsPath: path.resolve(dirname, source),
            })
          }),
        )
      }

      if (code !== undefined) {
        wxsMap.set(wxsPath, {
          emittedFile: {
            type: 'asset',
            fileName: configService.relativeAbsoluteSrcRoot(isRaw ? wxsPath : removeExtension(wxsPath)),
            source: code,
          },
        })
        wxsCodeCache.set(rawCode, code)
      }
    }
  }
  function handleWxsDeps(this: PluginContext, deps: WxmlDep[], absPath: string) {
    return Promise.all(
      deps.filter(x => x.tagName === 'wxs').map(async (wxsDep) => {
        const arr = wxsDep.value.match(/\.wxs(\.[jt]s)?$/)
        if (jsExtensions.includes(wxsDep.attrs.lang) || arr) {
          const wxsPath = path.resolve(path.dirname(absPath), wxsDep.value)

          await transformWxs.call(this, {
            wxsPath,
          })
        }
      }),
    )
  }

  return [
    {
      name: 'weapp-vite:wxs',
      enforce: 'pre',
      buildStart() {
        wxsMap.clear()
      },

      async buildEnd() {
        await Promise.all(
          Array.from(
            wxmlService
              .tokenMap
              .entries(),
          )
            .map(([id, token]) => {
              return handleWxsDeps.call(
                // @ts-ignore
                this,
                token.deps,
                id,
              )
            }),
        )

        Array.from(wxsMap.values()).forEach(({ emittedFile }) => {
          this.emitFile(emittedFile)
        })
      },
    },
  ]
}
