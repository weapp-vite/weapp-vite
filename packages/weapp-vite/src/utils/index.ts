export * from './file'
export * from './json'
export * from './md5'
export * from './projectConfig'
export * from './regexp'
export * from './version'
