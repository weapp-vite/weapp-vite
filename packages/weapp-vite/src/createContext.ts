import type { LoadConfigOptions } from './context'
import { getCompilerContext, resetCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const key = options?.key ?? 'default'
  if (!options?.key) {
    // ensure callers without explicit key do not reuse stale global context
    resetCompilerContext(key)
  }
  const ctx = getCompilerContext(key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
