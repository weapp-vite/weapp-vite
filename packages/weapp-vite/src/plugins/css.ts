import type { OutputAsset, OutputBundle } from 'rolldown'
import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '../context'
import type { SubPackageStyleEntry } from '../types'
import fs from 'fs-extra'
import path from 'pathe'
import { changeFileExtension, isJsOrTs } from '../utils'
import { cssCodeCache, processCssWithCache, renderSharedStyleEntry } from './css/shared/preprocessor'
import { collectSharedStyleEntries, injectSharedStyleImports, toPosixPath } from './css/shared/sharedStyles'

export { cssCodeCache }

async function handleBundleEntry(
  this: any,
  bundle: OutputBundle,
  bundleKey: string,
  asset: OutputAsset | OutputBundle[string],
  configService: CompilerContext['configService'],
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
) {
  if (asset.type !== 'asset') {
    return
  }

  const toAbsolute = (id: string) => {
    return path.isAbsolute(id) ? id : path.resolve(configService.cwd, id)
  }

  if (bundleKey.endsWith('.wxss')) {
    const [rawOriginal] = asset.originalFileNames ?? []
    const absOriginal = rawOriginal
      ? toAbsolute(rawOriginal)
      : path.resolve(configService.absoluteSrcRoot, bundleKey)
    const fileName = configService.relativeAbsoluteSrcRoot(absOriginal)

    if (fileName) {
      emitted.add(toPosixPath(fileName))
    }

    if (fileName && fileName !== bundleKey) {
      delete bundle[bundleKey]
      const css = await fs.readFile(absOriginal, 'utf8')
      this.emitFile({
        type: 'asset',
        fileName,
        source: css,
      })
    }

    return
  }

  if (!bundleKey.endsWith('.css')) {
    return
  }

  if (!asset.originalFileNames) {
    delete bundle[bundleKey]
    return
  }

  await Promise.all(
    asset.originalFileNames.map(async (originalFileName) => {
      if (!isJsOrTs(originalFileName)) {
        return
      }

      const modulePath = toAbsolute(originalFileName)
      const converted = changeFileExtension(modulePath, configService.outputExtensions.wxss)
      const fileName = configService.relativeAbsoluteSrcRoot(converted)
      if (!fileName) {
        return
      }
      const normalizedFileName = toPosixPath(fileName)
      const rawCss = asset.source.toString()
      const processedCss = await processCssWithCache(rawCss, configService)

      const cssWithImports = injectSharedStyleImports(
        processedCss,
        modulePath,
        fileName,
        sharedStyles,
        configService,
      )

      this.emitFile({
        type: 'asset',
        fileName,
        source: cssWithImports,
      })
      emitted.add(normalizedFileName)
    }),
  )

  delete bundle[bundleKey]
}

async function emitSharedStyleEntries(
  this: any,
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
  resolvedConfig?: ResolvedConfig,
) {
  if (!sharedStyles.size) {
    return
  }

  for (const entries of sharedStyles.values()) {
    for (const entry of entries) {
      const fileName = toPosixPath(entry.outputRelativePath)
      if (emitted.has(fileName)) {
        continue
      }

      const absolutePath = entry.absolutePath
      if (typeof this.addWatchFile === 'function') {
        this.addWatchFile(absolutePath)
      }

      if (!await fs.pathExists(absolutePath)) {
        continue
      }

      const { css: renderedCss, dependencies } = await renderSharedStyleEntry(entry, configService, resolvedConfig)
      if (typeof this.addWatchFile === 'function' && dependencies.length) {
        for (const dependency of dependencies) {
          if (dependency && dependency !== absolutePath) {
            this.addWatchFile(dependency)
          }
        }
      }

      const css = await processCssWithCache(renderedCss, configService)

      emitted.add(fileName)
      if (bundle[fileName]) {
        delete bundle[fileName]
      }

      this.emitFile({
        type: 'asset',
        fileName,
        source: css,
      })
    }
  }
}

async function emitSharedStyleImportsForChunks(
  this: any,
  sharedStyles: Map<string, SubPackageStyleEntry[]>,
  emitted: Set<string>,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
) {
  if (!sharedStyles.size) {
    return
  }

  const { outputExtensions } = configService

  await Promise.all(
    Object.values(bundle).map(async (output) => {
      if (output.type !== 'chunk') {
        return
      }

      const moduleId = output.facadeModuleId
      if (!moduleId) {
        return
      }

      const relativeModulePath = configService.relativeAbsoluteSrcRoot(moduleId)
      if (!relativeModulePath) {
        return
      }

      const converted = changeFileExtension(moduleId, outputExtensions.wxss)
      const fileName = configService.relativeAbsoluteSrcRoot(converted)
      if (!fileName) {
        return
      }

      const normalizedFileName = toPosixPath(fileName)
      if (emitted.has(normalizedFileName)) {
        return
      }

      const cssWithImports = injectSharedStyleImports(
        '',
        moduleId,
        fileName,
        sharedStyles,
        configService,
      )

      if (!cssWithImports.trim()) {
        return
      }

      const processedCss = await processCssWithCache(cssWithImports, configService)

      this.emitFile({
        type: 'asset',
        fileName,
        source: processedCss,
      })

      emitted.add(normalizedFileName)
    }),
  )
}

async function generateBundleSharedCss(
  this: any,
  ctx: CompilerContext,
  configService: CompilerContext['configService'],
  bundle: OutputBundle,
  resolvedConfig?: ResolvedConfig,
) {
  const sharedStyles = collectSharedStyleEntries(ctx, configService)
  const emitted = new Set<string>()
  const tasks = Object.entries(bundle).map(([bundleKey, asset]) => {
    return handleBundleEntry.call(this, bundle, bundleKey, asset, configService, sharedStyles, emitted)
  })

  await Promise.all(tasks)
  await emitSharedStyleEntries.call(this, sharedStyles, emitted, configService, bundle, resolvedConfig)
  await emitSharedStyleImportsForChunks.call(this, sharedStyles, emitted, configService, bundle)
}

export function css(ctx: CompilerContext): Plugin[] {
  const { configService } = ctx
  let resolvedConfig: ResolvedConfig | undefined
  return [
    {
      name: 'weapp-vite:css',
      enforce: 'pre',
      configResolved(config) {
        resolvedConfig = config
      },
      async generateBundle(_opts, bundle) {
        await generateBundleSharedCss.call(this, ctx, configService, bundle, resolvedConfig)
      },
    },
  ]
}
