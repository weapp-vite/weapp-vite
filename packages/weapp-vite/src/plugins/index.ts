import type { Plugin } from 'vite'
import type { CompilerContext } from '@/context'
import type { SubPackageMetaValue, WeappVitePluginApi } from '@/types'
import { wrapPlugin } from 'vite-plugin-performance'
import { asset } from './asset'
import { autoImport } from './autoImport'
import { autoRoutes } from './autoRoutes'
import { weappVite } from './core'
import { css } from './css'
import { preflight } from './preflight'
import { workers } from './workers'
import { wxs } from './wxs'

const RUNTIME_PLUGINS_SYMBOL = Symbol.for('weapp-runtime:plugins')

function attachRuntimePlugins(ctx: CompilerContext, plugins: Plugin[]): Plugin[] {
  const runtimePlugins = (ctx as any)[RUNTIME_PLUGINS_SYMBOL] as Plugin[] | undefined
  if (!runtimePlugins?.length) {
    return plugins
  }

  return [...runtimePlugins, ...plugins]
}

function applyInspect(ctx: CompilerContext, plugins: Plugin[]): Plugin[] {
  const inspectOptions = ctx.configService.weappViteConfig?.debug?.inspect
  if (!inspectOptions) {
    return plugins
  }

  // @ts-ignore third-party typings do not surface Plugin[] overload
  return wrapPlugin(plugins, inspectOptions)
}

function flatten(groups: Plugin[][]): Plugin[] {
  return groups.reduce<Plugin[]>((acc, cur) => {
    acc.push(...cur)
    return acc
  }, [])
}

export function vitePluginWeapp(
  ctx: CompilerContext,
  subPackageMeta?: SubPackageMetaValue,
): Plugin<WeappVitePluginApi>[] {
  const groups: Plugin[][] = [preflight(ctx)]

  if (!subPackageMeta) {
    groups.push(asset(ctx), autoRoutes(ctx), autoImport(ctx))
  }

  groups.push(weappVite(ctx, subPackageMeta), wxs(ctx), css(ctx))

  const assembled = attachRuntimePlugins(ctx, flatten(groups))
  if (subPackageMeta) {
    return assembled
  }

  return applyInspect(ctx, assembled)
}

export function vitePluginWeappWorkers(ctx: CompilerContext) {
  const groups = [preflight(ctx), workers(ctx)]
  const assembled = attachRuntimePlugins(ctx, flatten(groups))
  return applyInspect(ctx, assembled)
}
