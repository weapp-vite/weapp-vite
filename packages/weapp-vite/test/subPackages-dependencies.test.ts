import CI from 'ci-info'
import fs from 'fs-extra'
import path from 'pathe'
import { resetCompilerContext } from '@/context'
import { createCompilerContext } from '@/createContext'
import { cssCodeCache } from '@/plugins/css'
import { wxsCodeCache } from '@/plugins/wxs'
import { getFixture, scanFiles } from './utils'

describe.skipIf(CI.isCI)('subPackages-dependencies', () => {
  const cwd = getFixture('subPackages-dependencies')
  const distDir = path.resolve(cwd, 'dist')
  const nodeModulesDir = path.resolve(cwd, 'node_modules')

  beforeEach(async () => {
    await fs.remove(distDir)
    await fs.remove(nodeModulesDir)
  })

  afterEach(async () => {
    await fs.remove(distDir)
  })

  it('scanFiles', async () => {
    const ctx = await createCompilerContext({
      cwd,
    })
    try {
      await ctx.buildService.build()
      expect(await fs.exists(distDir)).toBe(true)
      const files = await scanFiles(distDir)
      expect(files).toMatchSnapshot()
      expect(wxsCodeCache.size).toBe(0)
      expect(cssCodeCache.size).toBe(4)

      expect(ctx.scanService.independentSubPackageMap).toMatchSnapshot()
      expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/buffer'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/gm-crypto'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/class-variance-authority'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm/buffer'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm/gm-crypto'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm/class-variance-authority'))).toBe(true)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm/clsx'))).toBe(false)
      expect(await fs.exists(path.resolve(distDir, 'packageB/miniprogram_npm/dayjs'))).toBe(false)
    }
    finally {
      ctx.watcherService.closeAll()
      resetCompilerContext()
    }
  })
})
