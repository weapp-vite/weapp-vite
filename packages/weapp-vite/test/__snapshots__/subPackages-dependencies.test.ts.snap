// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`subPackages-dependencies > scanFiles 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "miniprogram_npm/buffer/index.js",
  "miniprogram_npm/class-variance-authority/index.js",
  "miniprogram_npm/clsx/index.js",
  "miniprogram_npm/dayjs/index.js",
  "miniprogram_npm/gm-crypto/index.js",
  "packageA/index.js",
  "packageA/pages/cat.js",
  "packageA/pages/cat.json",
  "packageA/pages/cat.wxml",
  "packageA/pages/cat.wxss",
  "packageA/pages/dog.js",
  "packageA/pages/dog.wxml",
  "packageA/pages/dog.wxss",
  "packageB/index.js",
  "packageB/miniprogram_npm/buffer/index.js",
  "packageB/miniprogram_npm/class-variance-authority/index.js",
  "packageB/miniprogram_npm/gm-crypto/index.js",
  "packageB/pages/apple.js",
  "packageB/pages/apple.json",
  "packageB/pages/apple.wxml",
  "packageB/pages/banana.js",
  "packageB/pages/banana.wxml",
  "packageB/util.js",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "pages/index/index.wxss",
  "rolldown-runtime.js",
  "sitemap.json",
  "theme.json",
  "vendors.js",
]
`;

exports[`subPackages-dependencies > scanFiles 2`] = `
Map {
  "packageB" => {
    "entries": [
      "packageB/pages/apple",
      "packageB/pages/banana",
      "packageB/index",
    ],
    "subPackage": {
      "dependencies": [
        "class-variance-authority",
        "buffer",
        "gm-crypto",
      ],
      "entry": "index.js",
      "independent": true,
      "inlineConfig": undefined,
      "name": "pack2",
      "pages": [
        "pages/apple",
        "pages/banana",
      ],
      "root": "packageB",
    },
  },
}
`;
